### PR TITLE
Update for buffer API changes

### PR DESCRIPTION
**Pull request by jasongin**

### CODE REVIEW

**Comment by kkoopa** [comment]
The data argument of buffers should arguably be `void * `, using `char *` is a legacy oversight on Node's part. Using `char *` makes typecasting required for the user.

Even better would of course be to make them templated on the type of data.

**Comment by jasongin** [comment]
I made the Buffer class templated on the data type. For testing I updated my node-canvas NAPI port to use it, and it did indeed eliminate a lot of casting.

**Comment by jasongin** [comment]
Note there are some `reinterpret_cast<>` calls in the implementation that can go away after or change to `static_cast<>` after https://github.com/nodejs/abi-stable-node/issues/135 is fixed.

File: napi-inl.h [hunk 1]
```diff
@@ -975,41 +979,72 @@ inline Function::CallbackData::CallbackData(FunctionCallback cb, void* data)
...
 }
 
-inline char* Buffer::Data() const {
-  char* data;
-  napi_status status = napi_buffer_data(_env, _value, &data);
-  if (status != napi_ok) throw Error::New(Env());
-  return data;
+template <typename T>
+inline size_t Buffer<T>::Length() const {
+  if (_data == nullptr) {
```
Comment (boingoing) [inline napi-inl.h]: This same initialization code is duplicated in Buffer<T>::Length and Buffer<T>::Data. Would it make sense to put it into a helper like a private method Buffer<T>::EnsureCachedInfo or something to make the intent more clear and avoid having to change two places in the future? Also, might want to copy/paste this block again for another method later and inlining a small helper should translate to exactly the same code.

File: napi-inl.h [hunk 2]
```diff
@@ -979,55 +979,70 @@ inline Function::CallbackData::CallbackData(FunctionCallback cb, void* data)
...
+// Buffer<T> class
 ////////////////////////////////////////////////////////////////////////////////
 
-inline Buffer Buffer::New(Napi::Env env, size_t length) {
+template <typename T>
+inline Buffer<T> Buffer<T>::New(Napi::Env env, size_t length) {
   napi_value value;
   char* data;
-  napi_status status = napi_create_buffer(env, length, &data, &value);
+  napi_status status = napi_create_buffer(env, length * sizeof(T), &data, &value);
```
Comment (kkoopa) [inline napi-inl.h]: space after keyword, goes for the rest of the file
`sizeof(T) -> sizeof (T)`

File: napi-inl.h [hunk 3]
```diff
@@ -979,55 +979,70 @@ inline Function::CallbackData::CallbackData(FunctionCallback cb, void* data)
...
-inline size_t Buffer::Length() const {
+template <typename T>
+inline size_t Buffer<T>::Length() const {
   if (_data == nullptr) {
+    size_t byteLength;
     napi_status status = napi_get_buffer_info(
-      _env, _value, const_cast<char**>(&_data), const_cast<size_t*>(&_length));
+      _env, _value, reinterpret_cast<char**>(const_cast<T**>(&_data)), &byteLength);
     if (status != napi_ok) throw Error::New(Env());
+    *const_cast<size_t*>(&_length) = byteLength / sizeof(T);
```
Comment (kkoopa) [inline napi-inl.h]: Mark the field `_length` `mutable` instead.

File: napi-inl.h [hunk 4]
```diff
@@ -979,55 +979,70 @@ inline Function::CallbackData::CallbackData(FunctionCallback cb, void* data)
...
-inline char* Buffer::Data() const {
+template <typename T>
+inline T* Buffer<T>::Data() const {
   if (_data == nullptr) {
+    size_t byteLength;
     napi_status status = napi_get_buffer_info(
-      _env, _value, const_cast<char**>(&_data), const_cast<size_t*>(&_length));
+      _env, _value, reinterpret_cast<char**>(const_cast<T**>(&_data)), &byteLength);
     if (status != napi_ok) throw Error::New(Env());
+    *const_cast<size_t*>(&_length) = byteLength / sizeof(T);
```
Comment (kkoopa) [inline napi-inl.h]: Mark the field `_length` `mutable` instead.

File: napi-inl.h (hunk 1 above)
Comment (jasongin) [inline napi-inl.h]: Yes, I refactored this.

File: napi-inl.h (hunk 2 above)
Comment (jasongin) [inline napi-inl.h]: Fixed.

File: napi-inl.h (hunk 3 above)
Comment (jasongin) [inline napi-inl.h]: Fixed.

File: napi-inl.h (hunk 4 above)
Comment (jasongin) [inline napi-inl.h]: Fixed.

**Pull request merged**